appsrc/dcfile/dctable.man
.TH DCTABLE 1 "6 March 2014" "DICOM PS3" "DICOM PS3 - Make tables of attribute values"
.SH NAME
dctable \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Make tables of attribute values
.SH SYNOPSIS
.HP 10
.B dctable
" inputfile1 [ inputfile2 ... ]"
.so man1/gen.so
[
.B \-ignorereaderrors
]
[
.B \-v|verbose
]
[
.B \-r|recurse
]
[
.B \-describe
]
[
.B \-nofilename
]
[
.B \-noembeddedquotes|decimal
]
[
.B \-key|k " elementname|(0xgggg,0xeeee) ] ..."
[
.B \-input-nometa
]
[
.B \-input-ts " uid"
]
[
.B \-input-default
]
[
.B \-input-byteorder|-input-endian " big|little"
]
[
.B \-input-vr " implicit|explicit"
]
.SH DESCRIPTION
.LP
.B dctable
reads the named dicom input files and creates a tab delimited table of
the values of the selected attributes, with each value enclosed in double
quotes.
.LP
Binary attributes are written in hexadecimal with a preceding
"0x". Numeric string attributes are written in decimal.
.LP
The first column always contains the filenames.
.LP
A header is a added only if the -describe option is specified.
.LP
If a specified attribute is not present in a file, then it the corresponding
entry in the table will be empty (i.e. two consecutive quotes" but this
is not reported as an error.
.SH OPTIONS
The table of attribute values go to standard output, and the verbose
output and errors go to standard error.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-describe
.RS
Add a first line of output that is a header describing the attributes in each column.
.RE
.TP
.B \-nofilename
.RS
Do not include the filename in the table.
.RE
.TP
.B \-noembeddedquotes|decimal
.RS
Remove embedded quotes (") from within displayed values to allow them to be parsed consistenty later; also has the
side effect of displaying numeric binary values as decimal rather than hexadecimal, hence the alternative option name.
.RE
.TP
.B \-k|key " elementname|(0xgggg,0xeeee)"
.RS
Add a key to the list of attributes to be extracted. The attribute may also be specified
as a hexadecimal pair of the form (0xgggg,0xeeee), which also allows private attributes to be fetched.
.RE
.TP
.B \-r|recurse
.RS
If any attribute is not found in the top level dataset, recurse through sequence attribute items until
the first instance is found. This is a depth-first traversal, i.e., all children of the first item of
a sequence are searched before the sibling items.
.RE
.TP
.B \-ignorereaderrors
.RS
Attempt to extract key values even if errors encountered whilst parsing DICOM file.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% dctable -k Rows -k Columns
.RE
        -describe test.dc3
.RE
"Filename"	"Rows"	"Columns"
.RE
"test.dc3"	"0x0100"	"0x0100"
.RE
\ 
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcintro(1), dckey(1)
.SH AUTHOR
Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS
.LP
Attempts to extract sequences or OB,OW attributes may fail nastily.
.LP
The order in which the attributes are extracted and written is not the same
order as they are specified on the command line, and should not be depended
on.


CHANGES
241220:	libsrc/src/dctool/attrtypv.cc: Allow TAB character in control characters for text VRs (CP 1425)

241114:	appsrc/dcfile/dciodvfy.cc: Allow single byte of padding for odd number of 8 bit LUT entries in 16 bit words (000607)

241114:	appsrc/dcfile/dciodvfy.cc: Remove check on SegmentNumber monotonically increasing from 1 (not 0) if SegmentationType is LABELMAP (Sup 243)

241111:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ion Range Surface Compensator (CP 2401), Isodose levels for RT Dose visualization (CP 2402)

241105:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ocular Region Imaged Module in Ophthalmic Tomography (CP 2346)

241003:	libsrc/standard/condn.tpl,module/base.tpl: Restore not requiring SegmentationMacro if TILED_FULL (CP 2331) lost when adding Label Map Segmentation

240915:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add SOP Class and data elements for Height Map Segmentation and Revised Ophthalmic OCT En Face Image (Sup 240)

240914:	libsrc/standard/condn.tpl,sopcl.tpl,iodcomp,module,strval/base.tpl: Add Label Map Segmentation, and remove related experimental profile (Sup 243)

240913:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Clarify Description related to Range Shifter Settings reusing DICONDE Material Density (CP 2373 FT2)

240912:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Differentiate acquisition techniques for RT 3D CBCT imaging (CP 2220)

240822: libsrc/standard/module/base.tpl:	Correctly Add Ethnic Group Code Sequence (CP 2356)

240728: libsrc/standard/condn.tpl,module/base.tpl: Report error when PositionReferenceIndicator not SLIDE_CORNER for microscopy (000605)

240712:	appsrc/dcfile/dcacqmap.script,dcacqmap.man: Fix invalid character range in egrep for position files (000604), typo in documentation

240623:	libsrc/standard/elmdict/diconde.tpl: Add DICONDE data elements for thermography imaging

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Table Top Position Alignment UID to Table Top positions (CP 2376)

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Improve RT equipment mapping for imaging (CP 2374)

240611:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl:	Add Ethnic Group Code Sequence (CP 2356)

240610:	libsrc/standard/elmdict/dicom3.tpl:	Add data element for updated RT Patient Setup Module (CP 2344)

240515:	libsrc/standard/condn.tpl,module/base.tpl: Check ICCProfile and ColorSpace not present in top level dataset Image Pixel Module if Optical Path Module present (000603)

240506:	libsrc/standard/module/vl.tpl: Check AnnotationGroupNumber is not zero (000602)

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Detect FrameOfReference Module if FoR UID present even if position and orientation not, since MBPO in Secondary Capture IOD (CP 2330)

240402:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check AnnotationGroupUID is not duplicated (000601)

240329:	libsrc/standard/elmdict/other.tpl: 3DHISTECH private data elements from conformance statement

240329:	appsrc/dcfile/dcentvfy.cc,libsrc/src/locale/mesgtext.tpl: Check all ReferencedSOPInstanceUID instances are in set of supplied files (000600)

240328:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check sanity of Microscopy Bulk Simple Annotation indices and coordinates (000599)

240328:	libsrc/include/dctool/attr.h,attrtype.h,attrtypo.h,src/dctool/attrmxrd.cc,attrtypo.cc: Add support of in-memory (small) OL VR data elements to get values to check (000599)

240325:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Clinical Trial Identifiers need Issuers and may be multiple (CP2335)

240324:	libsrc/standard/condn.tpl,iodcomp,module/base.tpl: Clarify Segment scan order for TILED_FULL (CP 2331) and remove CP2331 profile option to test TILED_FULL segmentations

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Add Spatial Information to Secondary Capture IOD (CP 2330)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/vl.tpl: Add information about Vertex Distance to refraction correction parameters (CP 2324)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Add Synthetic Data Flag (CP 2320)

240316:	appsrc/dcfile/dcentvfy.cc: Check Frame of Reference IE attribute values same for same Series (000597)

240312:	libsrc/standard/elmdict/philips.tpl: Add more observed private data elements

240224:	appsrc/dcfile/dciodvfy.cc,libsrc/include/dctool/attr.h,attrothr.h,transyn.h,src/dctool/transyn.cc,transyne.h,locale/mesgtext.tpl,standard/transyn.tpl,support/transyn.awk: Add check that Basic Offset Table is valid if present, including length, point to Item Tag and begin with a JPEG/J2K SOI marker (000531) and point to even offsets (000526)

240222:	appsrc/acrnema/andump.cc,libsrc/include/dctool/attr.h,attrtype.h,attrtypn.h,attrtypo.h,dcstream.h,vr.h,generic/bnstream.h,src/dctool/attr.cc,attrmxrd.cc,attrnew.cc,attrtypo.cc: Add 64 bit binary VRs (CP 1819)

240215:	libsrc/standard/condn.tpl,module/vl.tpl: Correct spurious report of invalid CommonZCoordinateValue for nnotationCoordinateType not 3D when not actually present (000596)

240214:	appsrc/dcfile/dcentvfy.cc: Check Equipment IE attribute values same for same Series (000595)

240131:	libsrc/standard/condn.tpl,iodcomp,module,strval/base.tpl: Add profile option to test LABELMAP segmentations (Sup243)

240118:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check that SCOORD3D POLYGON are closed (first and last points same) (000592)

240114:	libsrc/standard/elmdict/dicom3.tpl: Provide additional ROI parameters (CP 2296)

240112:	libsrc/standard/elmdict/dicom3.tpl: Add new data elements for DICOMweb Storage Commitment Service (Sup 234)

231125:	config/Configure: Build two architectures when clang >=15 but not with ld flag "-force_cpusubtype_ALL" that is no longer supported

231115:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add Confocal Microscopy data elements and SOP Class UIDs (Sup 236)


appsrc/dcfile/dctable.cc
static const char *CopyrightIdentifier(void) { return "@(#)dctable.cc Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved."; }
#if USESTANDARDHEADERSWITHOUTEXTENSION == 1
#include <fstream>
#else
#include <fstream.h>
#endif

#if EMITUSINGSTDNAMESPACE == 1
using namespace std;
#endif


#include "attrmxls.h"
#include "attrnew.h"
#include "attrval.h"
#include "attrseq.h"
#include "mesgtext.h"
#include "dcopt.h"
#include "dcstream.h"
#include "elmconst.h"

static const char *options_output_key[] = {
	"key",
	"k",
	0
};

// Same in dckey.cc

static void dumpCodeSequenceAsString(SequenceAttribute *as,TextOutputStream &log) {
	AttributeList **itemLists;
	int nItems = as->getLists(&itemLists);
	for (int i=0; i<nItems; ++i) {
		AttributeList *thisItemList = itemLists[i];
		if (thisItemList) {
			Attribute *aCodeValue = (*thisItemList)[TagFromName(CodeValue)];
			Attribute *aCodingSchemeDesignator = (*thisItemList)[TagFromName(CodingSchemeDesignator)];
			Attribute *aCodeMeaning = (*thisItemList)[TagFromName(CodeMeaning)];
			if (aCodeValue || aCodingSchemeDesignator || aCodeMeaning) {
				const char *vCodeValue = AttributeValue(aCodeValue,"");
				const char *vCodingSchemeDesignator = AttributeValue(aCodingSchemeDesignator,"");
				const char *vCodeMeaning = AttributeValue(aCodeMeaning,"");
				log << "('"
				    << vCodeValue
				    << "','"
				    << vCodingSchemeDesignator
				    << "','"
				    << vCodeMeaning
				    << "')"
				;
			}
		}
	}
}

// the list will have already been checked before calling; check only the sequences within the list ...
static Attribute *
findFirstInstanceOfAttributeNestedAnywhereInList(Tag tag,AttributeList& list) {
//cerr << "findFirstInstanceOfAttributeNestedAnywhereInList(): Searching for Tag ("; writeZeroPaddedHexNumber(cerr,tag.getGroup(),4); cerr << ","; writeZeroPaddedHexNumber(cerr,tag.getElement(),4); cerr << ")" << endl;
	AttributeListIterator listi(list);
	while (!listi) {
		Attribute *a=listi();
//cerr << "findFirstInstanceOfAttributeNestedAnywhereInList(): Checking Attribute ("; writeZeroPaddedHexNumber(cerr,a->getTag().getGroup(),4); cerr << ","; writeZeroPaddedHexNumber(cerr,a->getTag().getElement(),4); cerr << ")" << endl;
		if (a->isSequence()) {
//cerr << "findFirstInstanceOfAttributeNestedAnywhereInList(): Checking within Sequence ("; writeZeroPaddedHexNumber(cerr,a->getTag().getGroup(),4); cerr << ","; writeZeroPaddedHexNumber(cerr,a->getTag().getElement(),4); cerr << ")" << endl;
			SequenceAttribute *as = (SequenceAttribute *)a;
			AttributeList **itemLists;
			int nItems = as->getLists(&itemLists);
			for (int i=0; i<nItems; ++i) {
				AttributeList *thisItemList = itemLists[i];
				if (thisItemList) {
					Attribute *aFound = (*thisItemList)[tag];
					if (!aFound) {
						 aFound = findFirstInstanceOfAttributeNestedAnywhereInList(tag,(*thisItemList));
					}
					if (aFound) {
//cerr << "findFirstInstanceOfAttributeNestedAnywhereInList(): Found ("; writeZeroPaddedHexNumber(cerr,aFound->getTag().getGroup(),4); cerr << ","; writeZeroPaddedHexNumber(cerr,aFound->getTag().getElement(),4); cerr << ")" << endl;
						return aFound;
					}
				}
			}
		}
		++listi;
	}
	return NULL;
}

static void
writeQuoteRemovedDecimalConvertedData(Attribute *a,TextOutputStream &stream)
{
	int n = a->getVM();
	for (int i=0; i<n; ++i) {
		if (i>0) {
			stream << "\\";
		}
		char *string;
		Uint32 l = strlen(string);
		if (a->getValue((Uint16)i,string)) {
			l = strlen(string);
		}
		else {
			l = 0;
		}
		const char *p = string;
		while (l--) {
			char c = *p++;
			if (c != '"') {
				stream << c;
			}
		}
	}
}			

int
main(int argc,char **argv)
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);

	bool verbose=options.get("verbose") || options.get("v");

	bool describe=options.get("describe");

	bool nofilename=options.get("nofilename");
	
	bool recurseToFindIt=options.get("recurse") || options.get("r");
	
	bool ignorereaderrors=options.get("ignorereaderrors");
	
	bool noembeddedquotes=options.get("noembeddedquotes") || options.get("decimal");

	bool bad=false;

	// identical to dckey.cc; should refactor this, perhaps as AttributeTag::read() ? :(
	ElementDictionary optdict;
	AttributeList *keylist=0;
	const char *keyarg;
	while (options.get(options_output_key,keyarg)) {
		Assert(keyarg);
		Tag tag;
		if (!getAttributeTagFromStringHexForm(keyarg,tag)	// handle "-k (0xgggg,0xeeee)" case
		 && !optdict.getTag(keyarg,tag)) {			// handle "-k keyword" case
			bad=true;
			cerr << "-" << options_output_key[0] << ": "
			     << EMsgDC(UnrecognizedElement)
			     << " - \"" << keyarg << "\""
			     << endl;
		}
		else {
			const char *vr=optdict.getValueRepresentation(tag);
			if (!vr) vr="UN";
			Attribute *a=newAttribute(vr,tag);
			Assert(a);
			if (!keylist) {
				keylist=new AttributeList;
				Assert(keylist);
			}
			(*keylist)+=a;
		}
	}

	dicom_input_options.done();
	Assert(!dicom_input_options.filename);

	int numberofinputfiles=!options;
	const char **listoffilenames = new const char * [numberofinputfiles];
	const char **ptr = listoffilenames;
	const char *filename;

	while(!options && (filename=options())) {
		++options;
		*ptr++=filename;
	}

	options.done();

	cerr << dicom_input_options.errors();
	cerr << options.errors();

	if (!dicom_input_options.good()
	 || !options.good()
	 || !options
	 || bad) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< " [-ignorereaderrors]"
			<< " [-describe]"
			<< " [-nofilename]"
			<< " [-noembeddedquotes|decimal]"
			<< " [-key|k elementname|(0xgggg,0xeeee)]"
			<< " [-recurse|r]"
			<< " [-verbose|v]"
			<< " " << MMsgDC(InputFile) << " ["<< MMsgDC(InputFile) << " ...]"
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		return 1;
	}
	
	if (!keylist) {	// (000482) and Debian 715809
		return 0;
	}

	bool success=true;
	TextOutputStream log(cerr);
	TextOutputStream outstream(cout);

	if (describe) {
		bool addtab=false;
		if (!nofilename) {
			outstream << "Filename";
			addtab=true;
		}
		AttributeListIterator listi(*keylist);
		while (!listi) {
			Attribute *ak=listi();
			Assert(ak);
			if (addtab) {
				outstream << "\t";
			}
			addtab=true;
			Tag tag=ak->getTag();
			const char *desc=optdict.getDescription(tag);
			if (desc) {
				outstream << "\"" << desc << "\"";
			}
			else {
				outstream << "(";
				writeZeroPaddedHexNumber(outstream,tag.getGroup(),4);
				outstream << ",";
				writeZeroPaddedHexNumber(outstream,tag.getElement(),4);
				outstream << ") " << flush;
			}
			++listi;
		}
		outstream << endl;
	}

	int i;
	for (i=0; i < numberofinputfiles; ++i) {
		const char *filename=listoffilenames[i];
		Assert(filename);
		if (verbose) log << "Reading \"" << filename << "\"" << endl;

#ifdef USEBINARYFLAGFORINPUTOPENMODE
		ifstream *fstr=new ifstream(filename,ios::in|ios::binary);
#else
		ifstream *fstr=new ifstream(filename);
#endif
		if (!fstr || !*fstr || !fstr->rdbuf()->is_open()) {
			log << AMsgDC(FileReadOpenFailed);
			if (filename) log <<" - \"" << filename << "\"";
			success=false;
			if (fstr) delete fstr;
			break;
		}

		DicomInputStream din(*(istream *)fstr,
			dicom_input_options.transfersyntaxuid,
			dicom_input_options.usemetaheader);

		ManagedAttributeList list;

		if (verbose) log << "******** While reading ... " << filename << " ... ********" << endl; 
		list.read(din,false/*newformat*/,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

		if (!list.good()) {
			log << list.errors()
			    << filename << ": " << EMsgDC(DatasetReadFailed) << endl;
			success=false;
			if (!ignorereaderrors) {
				if (fstr) delete fstr;
				break;
			}
		}
		{
			bool addtab=false;
			if (!nofilename) {
				outstream << "\"" << filename << "\"";
				addtab=true;
			}
			ElementDictionary *dict=list.getDictionary();
			Assert(dict);
			AttributeListIterator listi(*keylist);
			while (!listi) {
				Attribute *ak=listi();
				Assert(ak);
				if (addtab) {
					outstream << "\t";
				}
				addtab=true;
				Tag tag=ak->getTag();
				Attribute *ad=list[tag];
				if (!ad && recurseToFindIt) {
					ad = findFirstInstanceOfAttributeNestedAnywhereInList(tag,list);
				}
				if (ad) {
					outstream << "\"";
					if (ad->isSequence()) {
						SequenceAttribute *as = (SequenceAttribute *)ad;
						dumpCodeSequenceAsString(as,outstream);
					}
					else {
						if (noembeddedquotes) {
							writeQuoteRemovedDecimalConvertedData(ad,outstream);
						}
						else {
							ad->writeData(outstream);
						}
					}
					outstream << "\"";
				}
				else {
					outstream << "\"\"";
				}
				++listi;
			}
			outstream << endl;
		}

		if (fstr) delete fstr;
	}

	if (numberofinputfiles && listoffilenames) delete[] listoffilenames;

	return success ? 0 : 1;
}

